src/imars3d/backend/autoredux.py

```diff
@@ -79,26 +77,89 @@ def load_template_config(config_path: Union[str, Path]) -> dict:
+
     Parameters
     ----------
     data_file
-        The data file to extract information from.
-        ex. /HFIR/CG1D/IPTS-25777/raw/ct_scans/iron_man/20191029_ironman_small_0070_000_000_0002.tiff
+        The data file to extract information
 
     Returns
     -------
     dict
-        The extracted information.
+        The extracted information with the following keys:
+        facility, instrument, ipts, prepath, subpath
     """
-    extracted_data = data_file.split("/")
+    # Assuming the path is an absolute path
+    facility_position = 1
+    instrument_position = 2
+    experiment_position = 3
+    raw_position = 4
+
+    extracted_data = Path(data_file).parts
+    assert extracted_data[0] == os.path.sep, f"Path {str(data_file)} is not an absolute path"
+
+    # Extract FACILITY, INSTRUMENT, and IPTS
     data_dict = {}
-    # index from right to be agnostic of root
     data_dict["facility"] = extracted_data[facility_position]  # "HFIR"
     data_dict["instrument"] = extracted_data[instrument_position]  # "CG1D"
     data_dict["ipts"] = extracted_data[experiment_position]  # "IPTS-25777"
+
+    # Locate the index of the radiographs directory in list `extracted_data`
+    ct_position = -1
+    for i in range(1 + raw_position, len(extracted_data)):
+        if extracted_data[i] in ["ct_scans", "ct", "radiographs"]:
+            ct_position = i
+            break
+    assert ct_position > raw_position, "No radiographs directory found"
+    data_dict["radiograph"] = extracted_data[ct_position]
+
+    # Extract PREPATH and SUBPATH
+    data_dict["prepath"] = os.path.sep.join(extracted_data[1 + raw_position : ct_position])
+    data_dict["subpath"] = os.path.sep.join(extracted_data[1 + ct_position : -1])
+
     logger.info("Extract information from data file path.")
     return data_dict
+
+
+def substitute_template(config: dict, values: dict) -> dict:
+    r"""Update the template configuration with actual values
+
+    Parameters
+    ----------
+    config
+        dictionary resulting from loading the template JSON configuration file
+    values
+        dictionary resulting from scanning the path to the input radiograph
+        with function `extract_info_from_path`
+
+    Returns
+    -------
+        dictionary with template keywords substituted with actual values
+    """
+    # Compose ct_dir, ob_dir, and dc_dir as PREPATH/IMAGE_TYPE/SUBPATH
+    raw_path = Path("/")
+    for subdir in ("facility", "instrument", "ipts"):
+        raw_path = raw_path / values[subdir]
+    pre_path = raw_path / "raw" / values["prepath"]
+    values.update(
+        {
+            "ctdir": str(pre_path / values["radiograph"] / values["subpath"]),
+            "obdir": str(pre_path / "ob" / values["subpath"]),
+            "dcdir": str(pre_path / "df" / values["subpath"]),
+        }
+    )
+    template = Template(json.dumps(config))
+    return json.loads(template.substitute(**values))
```

Comment:
When I use this, it fails with a keyerror because values doesn't have every key that config does. using "template.safe_substitute(**values)" works for me though.

Reply:
what are the missing keys in `values`?

Reply:
"workingdir" and "outputdir" I believe.